### PR TITLE
Add support for GitHub Enterprise

### LINKED ISSUE #1129: Allow connection to GitHub Enterprise
The task `cci service connect github` does not allow a user to enter their GitHub enterprise server's address

### PR DESCRIPTION
**Pull request by davidjray**
Fixes #1129 

**Final Implementation Notes**
1. Split GitHub and GitHubEnterprise support into two distinct services. This maintains the historic GitHub service behavior, and adds support for GitHub Enterprise Server via a new service type.
2. In order to support internal Github Enterprise instances we must set an ENV variable `CUMULUSCI_SYSTEM_CERTS` which will allow our connection to support the SSL signed by the internal certificate authority. See [here](https://github.com/SFDO-Tooling/CumulusCI/blob/main/cumulusci/utils/http/requests_utils.py#L23).

Previous work notes:
1. ~~With this implementation the first GitHub service that have a non-github.com domain configured are selected, therefore, this implementation does not support multiple services under a given enterprise "domain", or respect the default service.~~ Fixed by #3271 and #3275
2. ~~The default github.com service is selected instead of the first matching service, which differs from the planned implementation, but this is better.~~ #3271 and #3275
3. ~~We need to refactor the token factory for github services to support the domain supplied when configuring a new service. This means the token must be manually supplied via `--token` when connecting a new github service.~~ Not needed due to #3271
4. ~~Supporting self signed certificates is the default for GitHub Enterprise connections, but this should be optional~~ Considered in #3287 but decided against in favor of setting CUMULUSCI_SYSTEM_CERTS

### CODE REVIEW

File: requirements/prod.in
Comment (jstvz) [inline requirements/prod.in:7]: Blocked by missing 3PP.

File: requirements/prod.txt
Comment (jstvz) [inline requirements/prod.txt:2]: ❔: Mind the version here, pip-tools is platform dependent IIRC.

File: cumulusci/core/github.py
Comment (jstvz) [inline cumulusci/core/github.py:47]: 💭: Needs discussion/review.

File: cumulusci/core/github.py
Comment (davidjray) [inline cumulusci/core/github.py:81]: Note to self... Do we sill like this name?

**Comment by davidjray** [comment]
[f6e5e05](https://github.com/SFDO-Tooling/CumulusCI/pull/3256/commits/f6e5e05fe3b8e822b303519b4d16e63b708be6f8) Fixes some issues I ran into when smoke testing. These fixes were required to overcome issues when configuring the first new "github_enterprise" service, and also access properties of services stored in the keychain.

File: cumulusci/core/github.py
Comment (davidjray) [reply cumulusci/core/github.py:81]: Yes the SSL verification will be done. This client, `GitHubEnterprise` has a `verify` flag that can be set to false, but we are not supporting that.

File: cumulusci/core/tests/test_github.py
Comment (davidjray) [reply cumulusci/core/tests/test_github.py:209]: These are NOT real credentials

**Review by jofsky — APPROVED** [review]
Can you ensure setuptools is not a part of prod.txt file?

File: cumulusci/core/github.py
Comment (jofsky) [reply cumulusci/core/github.py:164]: Can we do the default repo_domain setting for both services?

File: cumulusci/core/config/project_config.py
Comment (jstvz) [inline cumulusci/core/config/project_config.py:399]: Changing this function signature needs to be called out as a breaking change in release notes.

File: cumulusci/core/github.py
Comment (jstvz) [inline cumulusci/core/github.py:77]: ↔️ 🏌🏾‍♂️ 
```suggestion
    is_github: bool = host in (None, "None") or "github.com" in host
```

File: cumulusci/core/github.py
Comment (jstvz) [reply cumulusci/core/github.py:164]: > Can we do the default repo_domain setting for both services?

Would you say more about why you'd want that?

File: cumulusci/utils/git.py
Comment (jstvz) [inline cumulusci/utils/git.py:62]:
```suggestion
    """Parses a given Github URI into Owner, Repo Name, and Host
```

**Review by jstvz — APPROVED** [review]

File: cumulusci/cumulusci.yml
Comment (jstvz) [reply cumulusci/cumulusci.yml:1352]: @davidjray Blocking change requested.

**Review by jstvz — COMMENTED** [review]
LGTM

@jofsky I have code in this branch now, so I can't approve. Would you do so when ready?

**Comment by davidmreed** [comment]
@davidjray @jstvz @jofsky I think from above this branch is ready for merge, no? I can go ahead and approve and merge unless anyone has a concern.

**Comment by jstvz** [comment]
@davidmreed: 

> I think from above this branch is ready for merge, no? I can go ahead and approve and merge unless anyone has a concern.

Yeah, looks like it slipped the net when the WI was closed. Please approve.

**Review by prescod — APPROVED** [review]
Approving without review based on David Reed and James' attestation that they've reviewed.

**Pull request merged**
